File: assign2/haskell/src/Fibonacci.hs
module Fibonacci (fibonacciIterativeIterative, fibonacciRecursiveRecursive, fibonacciRecursiveIterative) where

fibonacciIterativeIterative :: Int -> [Int]
fibonacciIterativeIterative number = map fst $ scanl fibStep (1, 1) [1..number]
  where
    fibStep (curr, next) _ = (next, curr + next)

fibonacciRecursiveRecursive :: Int -> [Int]
fibonacciRecursiveRecursive 0 = [1]
fibonacciRecursiveRecursive 1 = [1, 1]
fibonacciRecursiveRecursive n = 
    let fibSequence = fibonacciRecursiveRecursive (n - 1)
    in fibSequence ++ [sum (drop (length fibSequence - 2) fibSequence)]

fibonacciRecursiveIterative :: Int -> [Int]
fibonacciRecursiveIterative n = fibHelper n [1, 1]
  where
    fibHelper 0 _ = [1]
    fibHelper 1 series = series
    fibHelper number series = fibHelper (number - 1) (series ++ [sum (drop (length series - 2) series)])

File: assign2/erlang/src/fibonacci.erl
-module(fibonacci).
-export([fibonacci_iterative_iterative/1, fibonacci_recursive_recursive/1, fibonacci_recursive_iterative/1]).

fibonacci_iterative_iterative(Number) when Number >= 0 ->
    case Number of
        0 -> [1];
        1 -> [1, 1];
        _ ->
            {Series, _, _} = lists:foldl(
                fun(_, {AccList, Prev, Curr}) ->
                    {AccList ++ [Prev + Curr], Curr, Prev + Curr}
                end,
                {[1, 1], 1, 1},
                lists:seq(2, Number)
            ),
            Series
    end.

fibonacci_recursive_recursive(0) -> [1];
fibonacci_recursive_recursive(1) -> [1, 1];
fibonacci_recursive_recursive(N) when N > 1 ->
    Series = fibonacci_recursive_recursive(N - 1),
    Series ++ [lists:last(Series) + lists:nth(N - 1, Series)].

fibonacci_recursive_iterative(N) when N >= 0 ->
    fibonacci_helper(N, [1, 1]).

fibonacci_helper(0, _) -> [1];
fibonacci_helper(1, Series) -> Series;
fibonacci_helper(N, Series) ->
    fibonacci_helper(N - 1, Series ++ [lists:nth(length(Series) - 1, Series) + lists:nth(length(Series), Series)]).
